documentation/proposals/Proposal - Silk.NET 3.X Generic Maths.md

```diff
@@ -0,0 +1,884 @@
+# Summary
+A new API for generic maths in Silk.NET.Maths 3.X, leveraging modern .NET features such as `INumber<T>` and vectorization.
+
+This API aims to replace the existing implementation of Silk.NET.Maths.
+
+# Contributors
+- Maxine H (@uwx)
+
+# Current Status
+- [x] Proposed
+- [ ] Discussed with API Review Board (ARB)
+- [ ] Approved
+- [ ] Implemented
+
+# Design Decisions
+- Should replace the existing APIs without too much growing pains
+- Should try to roughly match System.Numerics APIs, with concessions for design oversights in those APIs
+- Should use vectorized implementations for methods when possible
+- Prefer generic specialization and extension methods, and keep the type of `T` as wide as possible in the struct itself
+- Primitives are read-only, `with` preferred for modification
+- Public fields instead of properties
+- Implement the following interfaces:
+    - IEquatable
+    - IEquatable to equivalent System.Numerics primitive, if available
+    - IReadOnlyList
+    - IUtf8SpanFormattable
+    - IUtf8SpanParsable
+- Provide explicit implementations of the following interfaces:
+    - INumberBase (except for methods which are relevant to vectors)
+- Provide implementations of the following interfaces, but as extension/static methods:
+    - IHyperbolicFunctions
+    - ITrigonometricFunctions
+    - ILogarithmicFunctions
+    - IExponentialFunctions
+    - IPowerFunctions
+    - IRootFunctions
+    (The reasoning for this is that the required interface types are more narrow than INumberBase)
+- Static helper methods, where the result is expected to be bounded to 0-1 or the type of T has to be scoped to a type narrower than the one used by the primitive, have two variants:
+    - Method&lt;T>: T is scoped to a narrower type (e.g IRootFunctions&lt;T>, IFloatingPoint&lt;T>)
+    - Method&lt;T, TReturn>: no scope for T, but TReturn is scoped to a narrower type
```

Comment:
Framework Design Guidlines effectively recommend `T in, T out`. Conversions should be explicit.

Reply:
🤔 I think this may be one for the WG, I agree in principle but also we have an overload without the extra type parameter, so the only reason the extra one would be provided would be to convert. Of course, we could make this even more explicit, but we’ve always erred on the side of “most likely” user intent than 100% correct. This itself could be debated, and likewise could not do this for maths, I think it may make sense to discuss this in an open forum as these are philosophical decisions that should be made as a collective :) no hard disagree or hard agree here, which is why it’s interesting haha